app/services/auth.server.ts

```diff
@@ -0,0 +1,204 @@
+    const displayNames = [
+      ...new Set(apiUser.providers.flatMap((provider) => provider.displayNames)),
+    ]
+    const displayName = displayNames[0]
+
+    if (typeof decodedToken?.payload === 'string') {
+      throw new Error('Invalid token payload')
+    }
+
+    const expiresAt = new Date((decodedToken?.payload.exp ?? 0) * 1000)
+
+    let user: User = {
+      id: apiUser.info.id,
+      email,
+      displayName,
+      token: accessToken,
+      permissions: apiUser.info.permissions,
+      groups: apiUser.info.groups
+        .map((groupId) => {
+          return groups.find((group) => group.id === groupId)?.name
+        })
+        .filter(isTruthy),
+      expiresAt,
+    }
+
+    return user
+  } catch (err) {
+    console.error('Failed to generate user', err)
+    throw err
+  }
+}
+
+const verify: RfdVerifyCallback<User> = async ({ tokens }) => {
+  const accessToken = tokens.accessToken()
+  return fetchUser(accessToken)
+}
+
+const googleOAuth = new RfdOAuthStrategy(
+  {
+    host: process.env.RFD_API || '',
+    clientId: process.env.RFD_API_CLIENT_ID || '',
+    clientSecret: process.env.RFD_API_CLIENT_SECRET || '',
+    redirectURI: process.env.RFD_API_GOOGLE_CALLBACK_URL || '',
+    remoteProvider: 'google',
+    scopes: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  verify,
+)
+auth.use(googleOAuth)
+
+const githubOAuth = new RfdOAuthStrategy(
+  {
+    host: process.env.RFD_API || '',
+    clientId: process.env.RFD_API_CLIENT_ID || '',
+    clientSecret: process.env.RFD_API_CLIENT_SECRET || '',
+    redirectURI: process.env.RFD_API_GITHUB_CALLBACK_URL || '',
+    remoteProvider: 'github',
+    scopes: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  verify,
+)
+auth.use(githubOAuth)
+
+const magicLink = new RfdMagicLinkStrategy(
+  {
+    storage: sessionStorage,
+    host: process.env.RFD_API || '',
+    clientSecret: process.env.RFD_API_MLINK_SECRET || '',
+    pendingPath: '/login?email=sent',
+    returnPath: '/auth/magic/callback',
+    channel: 'login',
+    linkExpirationTime: 60 * 10, // 10 minutes (in seconds)
+    scope: ['group:info:r', 'rfd:content:r', 'rfd:discussion:r', 'search', 'user:info:r'],
+  },
+  async ({ token }) => {
+    return fetchUser(token)
+  },
+)
+auth.use(magicLink)
+
+export async function getUserFromSession(request: Request): Promise<User | null> {
+  const session = await sessionStorage.getSession(request.headers.get('Cookie'))
+  const user = session.get('user')
+  return user
+}
+
+export async function authenticate(request: Request) {
+  return await getUserFromSession(request)
+}
+
+export async function logout(request: Request, redirectTo: string) {
+  const session = await sessionStorage.getSession(request.headers.get('Cookie'))
+  session.unset('user')
+  throw redirect(redirectTo, {
+    headers: { 'Set-Cookie': await sessionStorage.commitSession(session) },
+  })
+}
```

Comment:
It amounts to the same thing unless there's some key in the session other than user, but this would be marginally more straightforward as `await sessionStorage.destroySession(session)`, like they do in the [docs](https://sergiodxa.github.io/remix-auth/#logout-the-user).

Reply:
Yeah, we use a general single session cookie for multiple purposes so there may be different values within it. Probably in a future state we should split these apart.